### PR TITLE
Replace expecty with clue

### PR DESCRIPTION
**Pull request by zainab-ali**
This PR replaces expecty with an munit-like `clue` function.

It uses macros to generate clues. Unlike munit, clues are stored locally to the expect call. This avoids the need for a global lock, and means tests can run concurrently, and add clues, without relying on any extra concurrency primitives.

It assumes that the code within an expect call is synchronous.

The behaviour is identical for Scala 2 and 3, but the macros are implemented differently.

Note that the signatures for `expect` aren't identical for Scala 2 and 3. The Scala 3 `expect` functions use context functions.

## Behaviour

A user can write:

```scala
val x = 1
val y = 2
expect(clue(x) == clue(y))
```

and will see the following on test failure:

```
Clues {
  x: Int = 1
  y: Int = 2
}
```

They can also nest calls to `clue`:

```scala
val x = 1
val y = 2
expect(clue(List(clue(x))) == clue(List(clue(y))))
```

## Scala 2 implementation

The Scala 2 implementation has a couple of macros:
 - The `ClueMacro` generates a clue for a value expression. This is identical to munit's clue macro.
 - The `ExpectMacro` constructs a local collection of `Clues`. Calls to the `clue` function are rewritten to call the `addClue` function of the local collection.
   
## Scala 3 implementation

In Scala 3, we can use context functions to provide the `Clues` collection. It still has a couple of macros:
 - The `ClueMacro` generates a clue for a value expression using simpler logic for obtaining the source code and type.
 - The `ExpectMacro` gets the source location to ensure that the tracing behaviour of `expect` is the same as it was previously. Calls to `clue` don't need to be rewritten.

### CODE REVIEW

File: modules/core/shared/src/main/scala-2/weaver/SourceLocationMacro.scala [hunk 1]
```diff
@@ -22,10 +22,10 @@ trait SourceLocationMacro {
 }
 
 object macros {
-  class Macros(val c: whitebox.Context) {
+  class Macros(val c: blackbox.Context) {
     import c.universe._
 
-    def fromContext: Tree = {
+    def fromContext: c.Tree = {
```
Comment (Baccata) [inline modules/core/shared/src/main/scala-2/weaver/SourceLocationMacro.scala:28]: 👍

File: modules/core/shared/src/main/scala-2/weaver/internals/ClueHelpers.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  /**
+   * Used to investigate failures in `expect` or `assert` statements.
+   *
+   * Surround a value with a call to `clue` to display it on failure.
+   */
+  @compileTimeOnly(
+    "This function can only be used within `expect` or `assert`.")
+  final def clue[A](@unused a: Clue[A]): A = {
+    // This function is removed as part of the `expect` macro expansion.
+    throw new Error("compileTimeOnly annotation not respected! This is likely to be a bug in weaver-test. Report it at https://github.com/typelevel/weaver-test/issues/new")
```
Comment (Baccata) [inline modules/core/shared/src/main/scala-2/weaver/internals/ClueHelpers.scala:17]: 👍

File: modules/core/shared/src/main/scala-2/weaver/ExpectMacro.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      cluesName: c.TermName,
+      value: c.Tree): c.Tree = {
+
+    import c.universe._
+    object transformer extends Transformer {
+
+      override def transform(input: Tree): Tree = input match {
+        case c.universe.Apply(fun, List(clueValue))
+            if fun.symbol == clueMethodSymbol =>
+          val transformedClueValue = super.transform(clueValue)
```
Comment (Baccata) [inline modules/core/shared/src/main/scala-2/weaver/ExpectMacro.scala:153]: I'm curious about this transformation here. Can you explain ?

**Comment by Baccata** [comment]
This is great ! Thank you so much for taking that on !🥇 

I just have a minor question, but I'm happy with the code, tests, and documentation.

File: modules/core/shared/src/main/scala-2/weaver/ExpectMacro.scala (hunk 3 above)
Comment (zainab-ali) [reply modules/core/shared/src/main/scala-2/weaver/ExpectMacro.scala:160]: Of course!

For context, the user writes `clue(x)` where `clue` refers to `ClueHelpers.clue`.   There's an implicit conversion from `x` to a `Clue`. Explicitly, this is written as `clue(Clue.generateClue(x))`  . 

The transformation replaces `ClueHelpers.clue` with `addClue` on a clues collection.  In other words, it replaces:
`clue(Clue.generateClue(x))` with `clues.addClue(Clue.generateClue(x))`.

The clues collection is generated as part of the overall `expect` macro. For example:

```scala
expect(clue(List(x, clue(y))))
```
becomes

```scala
expect {
  val clues = new Clues()
  clues.addClue(List(x, clues.addClue(y)))
}
```

**Review by Baccata — APPROVED** [review]

**Pull request merged**